NKL/Python.lean

```diff
@@ -0,0 +1,288 @@
+    set pos
+    let exp <- p key val
+    return (f exp pos)
+  | _ => throw "expecting object"
+
+private def withSrc (source : String) (p : Parser a) : Parser a :=
+  try set { lineno := 0 : Pos } ; p
+  catch e => get >>= throw ∘ genError e
+where
+  genError (err : String) (pos : Pos) : String :=
+    let lines := source.splitOn "\n"
+    let lineno := pos.lineno - 1
+    let colno := pos.col_offset
+    let line := if lines.length < lineno
+                then "<source not available>"
+                else lines[lineno]!
+    let indent := (Nat.repeat (List.cons ' ') colno List.nil).asString
+    s!"line {lineno}:\n{line}\n{indent}^-- {err}"
+
+-------------------------------------------------------------------------------
+-- Python AST Json objects
+
+def const : Json -> Parser Const
+  | .null => return .none
+  | .bool b => return (.bool b)
+  | .num jn => return (.num jn)
+  | .str "..." => return .ellipsis
+  | .str s => return (.string s)
+  | _ => throw "expecting constant"
+
+def exprCtx : Json -> Parser Ctx
+  | .str "Load" => return .load
+  | .str "Store" => return .store
+  | .str "Del" => return .del
+  | _ => throw "expecting ctx"
+
+partial def expr (j : Json) : Parser Expr :=
+  withPos expr' Expr.exprPos j
+where
+  expr' (key : String) (j : Json) : Parser Expr' := do
+    let strs := field (list str) j
+    let str := field str j
+    let ctx := field exprCtx j
+    let const := field const j
+    let exprs := field (list expr) j
+    let expr? := field (opt expr) j
+    let expr := field expr j
+    let keywords := field (list keyword) j
+    match key with
+    | "Constant" => return (.const (<- const "value"))
+    | "Name" => return (.name (<- str "id") (<- ctx "ctx"))
+    | "Attribute" => return (.attr (<- expr "value") (<- str "attr") (<- ctx "ctx"))
+    | "Tuple" => return (.tuple (<- exprs "elts") (<- ctx "ctx"))
+    | "List" => return (.list (<- exprs "elts") (<- ctx "ctx"))
+    | "Subscript" => return (.subscript (<- expr "value") (<- exprs "slice") (<- ctx "ctx"))
+    | "Slice" => return (.slice (<- expr? "lower") (<- expr? "upper") (<- expr? "step"))
+    | "BoolOp" => return (.boolOp (<- str "op") (<- exprs "values"))
+    | "BinOp" => return (.binOp (<- str "op") (<- expr "left") (<- expr "right"))
+    | "UnaryOp" => return (.unaryOp (<- str "op") (<- expr "operand"))
+    | "Compare" => return (.compare (<- expr "left") (<- strs "ops") (<- exprs "comparators"))
+    | "IfExp" => return (.ifExp (<- expr "test") (<- expr "body") (<- expr "orelse"))
+    | "Call" => return (.call (<- expr "func") (<- exprs "args") (<- keywords "keywords"))
+    | _ => throw s!"unsupported python construct {key}"
+
+  keyword (j: Json) : Parser Keyword := do
+    let j <- j.getObjVal? "keyword"
+    return ⟨ <- field str j "arg", <- field expr j "value", <- pos j ⟩
+
+partial def stmt (j : Json) : Parser Stmt :=
+  withPos stmt' Stmt.stmtPos j
+where
+  stmt' (key : String) (j : Json) : Parser Stmt' := do
```

Comment:
When should we prefer 

```
def stmt ...
  let stmt' ... = ...
```

vs

```
def stmt ...
  E
where
  ...
```
?

Reply:
Good question. I am not sure. I see both styles used in the Lean compiler without any clear rhyme or reason.